genie-web/src/main/java/com/netflix/genie/web/jpa/services/JpaClusterServiceImpl.java

```diff
@@ -590,6 +581,48 @@ public long deleteTerminatedClusters() {
         );
     }
 
+    private Map<Cluster, String> findClustersAndCommandsForJob(
+        final List<Set<String>> clusterCriteria,
+        final Set<String> commandCriterion
+    ) throws GenieServerException {
+        final Map<Cluster, String> foundClusters = Maps.newHashMap();
+        for (final Set<String> clusterCriterion : clusterCriteria) {
+            final List<Object[]> clusterCommands = this.clusterRepository.findClustersAndCommandsForCriterion(
```

Comment:
I think using a pair/tuple for this would be better than an `Object[2]`.
Type safe and can get rid of the next dozen or so lines.

Reply:
This might very well be true. I wrote it for `3.3` and since that native sql query is pretty complicated once I had it working I didn't really mess with it. If you look this method is just a copy of a code blocked moved to a helper method for re-use between the 4.0 API and the 3.x API. I'm open to improving it but I don't think we should do it as part of this PR particularly as once we change how the `Criterion` are handled for 4.x much of this may change anyway.